DD-1532: more unit tests and fixes

jo-pol opened the pull request:
Fixes DD-1532: more unit tests and fixes

# Description of changes

* third tranche of unit test
* Constructor of LayerManagerImpl checks for garbage in the stagingDir
* delete directories and file correctly from DB

# How to test


# Related PRs 
(Add links)
* 

# Notify
@DANS-KNAW/core-systems

jo-pol committed:
complete coverage of LayerManager

jo-pol committed:
no sneakyThrows, ignore garbage in staging-root

jo-pol committed:
fix getting the top layer

jo-pol committed:
tested getLayer

jo-pol committed:
archived status of getLayer results

jo-pol committed:
coverage of ZipArchive

jo-pol committed:
throw on garbage in stagingDir

jo-pol committed:
fix github glitch

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Test
+    public void should_use_the_existing_directory() throws IOException {
+        Files.createDirectories(stagingDir.resolve("1234567890123"));
```
Safer to use a local long type var and use that on both lines 38 and 42. It also makes it more explicit that this is the same value. Also modify this in the other tests.

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void should_throw_on_a_plain_file() throws IOException {
```
```suggestion
    public void should_throw_on_a_regular_file_in_the_staging_root() throws IOException {
```

That is what it is mostly called. "Plain" could cause confusion with "plain text".

janvanmansum commented on src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java:
```diff
@@ -40,37 +38,42 @@ public class LayerManagerImpl implements LayerManager {
-        try (Stream<Path> pathStream = Files.list(stagingRoot)) {
-            paths = pathStream.toList();
+    private void initTopLayer() throws IOException {
+        if (Files.notExists(stagingRoot)) {
```
Not sure about whether auto-creating the staging root can cause surprises, but I can't think of a concrete scenario, so I'll go along with your change.

janvanmansum commented on src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java:
```diff
@@ -40,37 +38,42 @@ public class LayerManagerImpl implements LayerManager {
+            throw new IllegalStateException("Not a directory: " + path);
+        }
+        if(!path.getFileName().toString().matches("\\d{13,}")) {
+            // more than 13 digits in nov 2286
```
Our Vault is forever :)

janvanmansum commented on src/main/java/nl/knaw/dans/layerstore/LayerManagerImpl.java:
```diff
@@ -100,7 +103,11 @@ private void archive(Layer layer) {
     public Layer getLayer(long id) {
-        if (stagingRoot.resolve(Long.toString(id)).toFile().exists() || archiveProvider.exists(Long.toString(id))) {
+        if (id == topLayer.getId()) {
+            // an empty layer would not be found in the staging root
```
I don´t understand this comment. Wouldn´t there be an empty staging directory?

Returning the `topLayer` here is OK.

jo-pol replied:
A new layer in the staging directory is only created when something is added. Adjusted the comment.

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void should_use_the_existing_directory() throws IOException {
```
```suggestion
    public void should_use_the_existing_staging_directory() throws IOException {
```

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void should_throw_on_the_too_short_directory_name() throws IOException {
```
```suggestion
    public void should_throw_on_staging_directory_name_that_is_too_short_to_be_a_recent_timestamp() throws IOException {
```

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void should_throw_on_an_alphanumeric_directory() throws IOException {
```
```suggestion
    public void should_throw_on_staging_directory_name_that_contains_letters() throws IOException {
```

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerConstructorTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void should_throw_on_a_directory_with_a_dot_in_the_name() throws IOException {
```
```suggestion
    public void should_throw_on_staging_directory_name_that_contains_decimal_point() throws IOException {
```

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java:
```diff
@@ -15,5 +15,120 @@
+    public void should_find_garbage_in_stagingDir_created_after_creating_LayerManagerImpl_object() throws IOException {
+        // Given
+        var layerManager = new LayerManagerImpl(stagingDir, new ZipArchiveProvider(archiveDir));
+        assertThat(layerManager.getTopLayer().getId()).isNotEqualTo(0L);
```
Why this "assert"? Also, it is under the "Given" comment.

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java:
```diff
@@ -15,5 +15,120 @@
+        var layer = layerManager.getLayer(0L);
+
+        // Then
+        assertThat(layer.getId()).isNotEqualTo(layerManager.getTopLayer().getId());
```
I also don't understand the point of the test.

jo-pol replied:
It shows inconsistent behavior between the constructor and getLayer.

janvanmansum commented on src/test/java/nl/knaw/dans/layerstore/LayerManagerGetLayerTest.java:
```diff
@@ -15,5 +15,120 @@
+        assertThat(archiveDir).isEmptyDirectory(); // nothing archived yet
+        sleep(1L); // to make sure to get a layer with another time stamp
+        layerManager.newTopLayer();
+        AssertionsForClassTypes.assertThat(initialLayerId).isNotEqualTo(layerManager.getTopLayer().getId());
```
Use static import.

janvanmansum reviewed (changes requested):
See inline comments. Mainly some i's to dot, but also one test that is unclear to me.

jo-pol committed:
lower level checks

jo-pol committed:
Apply suggestions from code review

Co-authored-by: Jan van Mansum <jan.van.mansum@dans.knaw.nl>

jo-pol committed:
extracted vars

https://github.com/DANS-KNAW/dans-layer-store-lib/pull/7#discussion_r1570744355

jo-pol committed:
more review comments

jo-pol committed:
Merge branch 'master' into DD-1532-3e

jo-pol committed:
fix stack trace

jo-pol committed:
moved the fix

jo-pol committed:
fixed delete dir from DB, not only children

jo-pol committed:
fixed delete file from DB, not only from storage
